Source/Urho3D/IO/PackageBuilder.cpp

```diff
@@ -0,0 +1,338 @@
+PackageBuilder::PackageBuilder():
+    buffer_(nullptr)
+{
+}
+
+bool PackageBuilder::WriteHeader()
+{
+    // Write ID, number of files & placeholder for checksum
+    if (!compress_)
+        buffer_->WriteFileID("RPAK");
+    else
+        buffer_->WriteFileID("RLZ4");
+
+    // Num files
+    if (!buffer_->WriteUInt(entries_.size()))
+        return false;
+    // Checksum
+    if (!buffer_->WriteUInt(checksum_))
+        return false;
+    // Version
+    if (!buffer_->WriteUInt(0))
+        return false;
+    // fileListOffset
+    if (!buffer_->WriteUInt64(fileListOffset_))
+        return false;
+
+    return true;
+}
+
+bool PackageBuilder::Create(AbstractFile* dest, bool compress)
+{
+    if (buffer_)
+    {
+        URHO3D_LOGERROR("Target file already set");
+        return false;
+    }
+
+    auto file = dynamic_cast<File*>(dest);
+    if (!file->IsOpen())
+    {
+        URHO3D_LOGERROR("Can't open file " + dest->GetName());
+        return false;
+    }
+
+    entries_.clear();
+    headerPosition_ = dest->GetPosition();
+    compress_ = compress;
+    buffer_ = dest;
+    file_ = dynamic_cast<RefCounted*>(dest);
+    return WriteHeader();
+}
+/// Append entry to package.
+bool PackageBuilder::Append(const ea::string& name, const SharedPtr<File>& file)
+{
+    return AppendImpl(name, file.Get());
+}
+
+/// Append entry to package.
+bool PackageBuilder::Append(const ea::string& name, MemoryBuffer& content)
+{
+    return AppendImpl(name, &content);
+}
+
+/// Append entry to package.
+bool  PackageBuilder::Append(const ea::string& name, const ByteVector& content)
+{
+    MemoryBuffer buffer(content);
+    return Append(name, buffer);
+}
+
+/// Append entry to package.
+bool PackageBuilder::AppendImpl(const ea::string& name, AbstractFile* content)
+{
+    if (!buffer_)
+    {
+        URHO3D_LOGERROR("PackageBuilder is not initialized");
+        return false;
+    }
+    auto& entry = entries_.emplace_back();
+    entry.name_ = name;
+    entry.offset_ = buffer_->GetPosition();
+    entry.checksum_ = 0;
+    entry.size_ = content->GetSize();
+
+    unsigned dataSize = content->GetSize();
+    ea::unique_ptr<unsigned char[]> buffer(new unsigned char[dataSize]);
+
+    if (content->Read(&buffer[0], dataSize) != dataSize)
+    {
+        URHO3D_LOGERROR("Could not read file " + buffer_->GetName());
+        return false;
+    }
+
+    for (unsigned j = 0; j < dataSize; ++j)
+    {
+        checksum_ = SDBMHash(checksum_, buffer[j]);
+        entry.checksum_ = SDBMHash(entry.checksum_, buffer[j]);
+    }
+
+    if (!compress_)
+    {
+        //if (!quiet_)
+        //    PrintLine(entries_[i].name_ + " size " + ea::to_string(dataSize));
+        buffer_->Write(&buffer[0], entry.size_);
+    }
+    else
+    {
+        CompressedStreamSerializer serializer(*buffer_);
+        constexpr auto blockSize_ = COMPRESSED_BLOCK_SIZE;
+        ea::unique_ptr<unsigned char[]> compressBuffer(new unsigned char[LZ4_compressBound(blockSize_)]);
```

Comment:
☝🏻

Reply:
What's about it? I've seen ea::unique_ptr used with arrays before in Urho code, I though this is how Urho rolls

Reply:
I see, some leftovers...